rpi 4.14.y: fkms support

anholt opened the pull request:
The forward port of the fkms code missed that enable/disable_vblank moved to the CRTC funcs, so we had fkms code in vc4_crtc.c that never got called because our enable/disable were entirely missing.  Also brings forward the T tiling patch that got dropped.

anholt committed:
drm/vc4: Add FB modifier support to firmwarekms.

Signed-off-by: Eric Anholt <eric@anholt.net>
(cherry picked from commit 11752d73488e08aaeb65fe8289a9c016acde26c2)

anholt committed:
drm/vc4: Add missing enable/disable vblank handlers in fkms.

Fixes hang at boot in 4.14.

Signed-off-by: Eric Anholt <eric@anholt.net>

pelwell commented:
With these commits I do now get a display, but there are still a few problems:
1. The fkms driver has terrible mouse performance - there's far too much buffering in the system.
2. It still gets the warning at line 1550 in drm_handle_vblank.
3. Both versions of the driver (or the overlay?) kill `sudo vcdbg log msg`, causing it to print a slew of:
```
debug_sym: vc_mem_copy: Failed to get memory size via ioctl: Device or resource busy(16)

debug_sym: vc_mem_copy: Failed to get memory size via ioctl: Device or resource busy(16)
```

popcornmix commented:
I've tried too. 3D performance is fine - neverball runs smoothly with fkms.
But the mouse is unusable - one to two seconds of lag on it.
This doesn't occur on 4.9 kernel. Same bad behaviour on both 4.14 and 4.15.

anholt committed:
drm/vc4: Fix warning about vblank interrupts before DRM core is ready.

The SMICS interrupt fires continuously, but since it's 1/100 the rate
of the USB interrupts, we don't really need a way to turn it off.  We
do need to make sure that we don't tell DRM about it until DRM has
asked for the interrupt at least once, because otherwise it will throw
a warning at boot time.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/vc4: Skip SET_CURSOR_INFO when the cursor contents didn't change.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/vc4: Remove duplicate primary/cursor fields from FKMS driver.

The CRTC has those fields and we can just use them.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt committed:
drm/vc4: Don't wait for vblank on fkms cursor updates.

We don't use the same async update path between fkms and normal kms,
and the normal kms workaround ended up making us wait.  This became a
larger problem in rpi-4.14.y, as the USB HID update rate throttling
got (accidentally?) dropped.

Signed-off-by: Eric Anholt <eric@anholt.net>

anholt commented:
mouse lag fixed, even without the HID throttling (~10% CPU in Xorg when moving the mouse quickly).  no idea on vcdbg log, which I think should be unrelated to this series..

pelwell commented:
If the rest is OK them I'm happy to shelve the vcdbg issue pending further investigation.

popcornmix commented:
I think the vcdbg issue is related to using dma from the fb driver to read the top 16M of sdram (as arm cannot see it). I think that gets disabled when fkms is enabled.
